src/passes/OptimizeInstructions.cpp

```diff
@@ -1615,6 +1585,170 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  Expression* partialEvaluation(Binary* binary) {
+    assert(binary->right->is<Const>());
+    Const* right = binary->right->cast<Const>();
+
+    // detect overflow or zero during signed multiplication
+    auto hasSignedMulOverflow = [](auto a, auto b) {
+      using T = decltype(a);
+      auto min = std::numeric_limits<T>::min();
+      auto max = std::numeric_limits<T>::max();
+      return (a != T(0) && b != T(0) &&
+              ((b > T(0) && a > max / b) || (b < T(0) && a < max / b) ||
+               (b > T(0) && a < min / b) || (b < T(-1) && a > min / b)));
+    };
+
+    // detect overflow or zero during signed multiplication
+    auto hasUnsignedMulOverflow = [](auto a, auto b) {
+      using T = decltype(a);
+      return a != T(0) && b != T(0) && a > std::numeric_limits<T>::max() / b;
+    };
+
+    // the square of some operations can be merged
+    // (x op C1) op C2  ==>  (x op (C1 op C2))
+    if (auto* left = binary->left->dynCast<Binary>()) {
```

Comment:
It would be great to use a matcher for at least the basic structural matching here, especially since this is a brand new function.